### PR TITLE
[FrameworkBundle][Notifier] Allow to configure or disable the message bus to use

### PR DESCRIPTION
**Pull request by jschaedl**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | - 
| License       | MIT
| Doc PR        | tbd 

For the mailer integration there is already a possibility to disable or configure the `message_bus` to use. See: https://github.com/symfony/symfony/issues/34633 and https://github.com/symfony/symfony/pull/34648

This PR introduces a similar config option `notifier.message_bus` for the notifier integration.

Things done:

- [x] Add a config option `notifier.message_bus` for the notifier configuration
- [x] Adjust FrameworkExtension considering the new `message_bus` config option
- [x] Add tests for the new config option

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier.php [hunk 1]
```diff
@@ -46,15 +46,24 @@
             ->tag('notifier.channel', ['channel' => 'browser'])
 
         ->set('notifier.channel.chat', ChatChannel::class)
-            ->args([service('chatter.transports'), service('messenger.default_bus')->ignoreOnInvalid()])
+            ->args([
+                service('chatter.transports'),
+                abstract_arg('message bus'),
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier.php:57]: Is it correct with the whitespace?

Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier.php:57]: Yes, that's correct. This is the description of the abstract argument.

**Comment by OskarStark** [comment]
It looks like your commit is not associated with your GitHub account:
image

And FrameworkBundle is twice in the PR header 🧐

**Comment by OskarStark** [comment]
Anything left todo here @jschaedl ?

**Comment by jschaedl** [comment]
Hi @OskarStark 
Implementation is done. But I have no idea why tests are failing on PHP 7.4 🤔

**Comment by OskarStark** [comment]
Can you please rebase and see if it's still failing ? Thanks

**Comment by OskarStark** [comment]
@jschaedl can you please check the failing tests?

Thanks

**Comment by jschaedl** [comment]
@OskarStark I rebased the branch. Tests are still failing on PHP 8.0 but this seems to be unrelated with my changes.

**Comment by nicolas-grekas** [comment]
The failures are related I believe.

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -2398,6 +2410,9 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
             GitterTransportFactory::class => 'notifier.transport_factory.gitter',
             ClickatellTransportFactory::class => 'notifier.transport_factory.clickatell',
             MicrosoftTeamsTransportFactory::class => 'notifier.transport_factory.microsoftteams',
+            LightSmsTransportFactory::class => 'notifier.transport_factory.lightsms',
+            SmsBiurasTransportFactory::class => 'notifier.transport_factory.smsbiuras',
+            MessageBirdTransport::class => 'notifier.transport_factory.messagebird',
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Should be in a separate PR as bugfix on `5.x`, thanks

Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: https://github.com/symfony/symfony/pull/40842

**Comment by jschaedl** [comment]
Needs rebase after https://github.com/symfony/symfony/pull/40842 was merged

**Comment by jschaedl** [comment]
@OskarStark I am about to split this PR. 
1. Adding the missing TransportFactories to the FrameworkBundle: https://github.com/symfony/symfony/pull/40843 ✅ 
2. I would like to add basic tests for the Framework integration for the Notifier: https://github.com/symfony/symfony/pull/40844 ✅ 
2. Adding a test to make sure we won't forget to add TransportFactories anymore: https://github.com/symfony/symfony/pull/40845 ✅ 
3. And then adding the possibility to add a specific message bus with this PR ✅

**Comment by jschaedl** [comment]
Rebased and ready for another review :-)

**Comment by OskarStark** [comment]
Ready to merge @nicolas-grekas ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 3]
```diff
@@ -2390,6 +2390,22 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
...
+            foreach ($servicesWithBusArgument as $serviceId) {
+                if ($container->hasDefinition($serviceId)) {
+                    $container->getDefinition($serviceId)->replaceArgument(1, null);
+                }
+            }
+        } else {
+            foreach ($servicesWithBusArgument as $serviceId) {
+                if ($container->hasDefinition($serviceId)) {
+                    $container->getDefinition($serviceId)->setArgument(0, null);
+                    $container->getDefinition($serviceId)->replaceArgument(1, $messageBus ? new Reference($messageBus) : new Reference('messenger.default_bus', ContainerInterface::NULL_ON_INVALID_REFERENCE));
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]:
```suggestion
                    $container->getDefinition($serviceId)
                        ->setArgument(0, null)
                        ->replaceArgument(1, $messageBus ? new Reference($messageBus) : new Reference('messenger.default_bus', ContainerInterface::NULL_ON_INVALID_REFERENCE))
                    ;
```

?

**Review by ogizanagi — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 3 above)
Comment (OskarStark) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]:
```diff
- ->setArgument(0, null);
+ ->setArgument(0, null)
```
?

Comment (ogizanagi) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: @OskarStark : Right, fixed the suggestion

Comment (OskarStark) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: friendly ping @jschaedl

**Comment by OskarStark** [comment]
Please rebase to get rid of Travis, thanks

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 3 above)
Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Thanks @ogizanagi and @OskarStark

**Comment by jschaedl** [comment]
@OskarStark Rebase done.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 4]
```diff
@@ -2390,6 +2390,24 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
...
+                if ($container->hasDefinition($serviceId)) {
+                    $container->getDefinition($serviceId)->replaceArgument(1, null);
+                }
+            }
+        } else {
+            foreach ($servicesWithBusArgument as $serviceId) {
+                if ($container->hasDefinition($serviceId)) {
+                    $container->getDefinition($serviceId)
+                        ->setArgument(0, null)
+                        ->replaceArgument(1, $messageBus ? new Reference($messageBus) : new Reference('messenger.default_bus', ContainerInterface::NULL_ON_INVALID_REFERENCE))
```
Comment (jderusse) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `$messageBus ? ...` can be removed because we are in the `else` part.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 5]
```diff
@@ -2390,6 +2390,24 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
             $container->removeDefinition('notifier.channel.email');
         }
 
+        $servicesWithBusArgument = ['texter', 'chatter', 'notifier.channel.chat', 'notifier.channel.email', 'notifier.channel.sms'];
+        if (false === $messageBus = $config['message_bus']) {
+            foreach ($servicesWithBusArgument as $serviceId) {
+                if ($container->hasDefinition($serviceId)) {
+                    $container->getDefinition($serviceId)->replaceArgument(1, null);
```
Comment (jderusse) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: shouldn't it be the `messenger.default_bus` service? like previous implementation

Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: This case handles the situation where I don't want to have a message bus at all.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 4 above)
Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:2405]: The idea was to define your own message bus. So if  $messageBus is not false, it contains a service id.

Comment (jschaedl) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: @jderusse  You are right. Will change it: 
```suggestion
                        ->replaceArgument(1, new Reference($messageBus, ContainerInterface::NULL_ON_INVALID_REFERENCE))
```

**Comment by chalasr** [comment]
Some tests are failing

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 6]
```diff
@@ -2390,6 +2390,24 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
             $container->removeDefinition('notifier.channel.email');
         }
 
+        $servicesWithBusArgument = ['texter', 'chatter', 'notifier.channel.chat', 'notifier.channel.email', 'notifier.channel.sms'];
+        if (false === $messageBus = $config['message_bus']) {
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: should we throw an exception if one tries to configure `message_bus` while messenger is not installed/enabled?

**Comment by fabpot** [comment]
@jschaedl Can you have a look at the failing tests?

**Comment by fabpot** [comment]
@jschaedl I've rebased on current 6.2 and fixed the tests. Can you review my commit before I merge?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 7]
```diff
@@ -2501,21 +2501,15 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
...
-                }
+        foreach (['texter', 'chatter', 'notifier.channel.chat', 'notifier.channel.email', 'notifier.channel.sms'] as $serviceId) {
+            if (!$container->hasDefinition($serviceId)) {
+                continue;
             }
-        } else {
-            foreach ($servicesWithBusArgument as $serviceId) {
-                if ($container->hasDefinition($serviceId)) {
-                    $container->getDefinition($serviceId)
-                        ->setArgument(0, null)
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I don't get why you are removing the transports here. We do need the transports in all cases.

**Comment by nicolas-grekas** [comment]
Friendly ping @jschaedl

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 6 above)
Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: The logic is used [here](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L2445) for setting the messageBus argument on the mailer.mailer service in the same way. I thought using `ContainerInterface::NULL_ON_INVALID_REFERENCE` is enough?

Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: - [ ] throw exception if messenger is not installed

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 7 above)
Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Not sure what you mean, but I am not removing the transports. I am adding the message bus service to every service with a messageBus argument, in case the `message_bus` config is filled.

**Comment by jschaedl** [comment]
I rebased again. Looks good to me now.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jschaedl.
